src/mapping/recurrences/finite_state_machine.jl

```diff
@@ -0,0 +1,275 @@
+            bsn_nfo.consecutive_match += 1
+            store_once_per_cell || store_attractor!(bsn_nfo, u)
+        elseif iseven(ic_label) && bsn_nfo.consecutive_match ≥ mx_chk_loc_att
+            # We have recorded the attractor with sufficient accuracy.
+            # We now set the empty counters for the new attractor.
+            cleanup_visited_cells!(bsn_nfo)
+            bsn_nfo.visited_cell_label += 2
+            bsn_nfo.current_att_label += 2
+            reset_basins_counters!(bsn_nfo)
+            # We return the label corresponding to the *basin* of the attractor
+            return ic_label + 1
+        end
+        return 0
+    end
+
+    # We've hit a cell labelled as a basin of an existing attractor.
+    # Note, this clause can only trigger if the basin array is NOT sparse.
+    if bsn_nfo.state == :bas_hit
+        if bsn_nfo.prev_label == ic_label
+            bsn_nfo.consecutive_match += 1
+        else
+            bsn_nfo.consecutive_match = 1
+        end
+        if  bsn_nfo.consecutive_match > mx_chk_hit_bas
+            cleanup_visited_cells!(bsn_nfo)
+            reset_basins_counters!(bsn_nfo)
+            return ic_label
+        end
+        bsn_nfo.prev_label = ic_label
+        return 0
+    end
+
+    # This state occurs when the dynamical system state is outside the grid
+    if bsn_nfo.state == :lost
+        bsn_nfo.consecutive_lost += 1
+        if bsn_nfo.consecutive_lost > mx_chk_lost || norm(u) > horizon_limit
+            cleanup_visited_cells!(bsn_nfo)
+            reset_basins_counters!(bsn_nfo)
+            # problematic IC: diverges or wanders outside the defined grid
+            return -1
+        end
+        bsn_nfo.prev_label = ic_label
+        return 0
+    end
+end
+
+function store_attractor!(bsn_nfo::BasinsInfo{D, G, Δ, T}, u) where {D, G, Δ, T}
+    # bsn_nfo.current_att_label is the number of the attractor multiplied by two
+    attractor_id = bsn_nfo.current_att_label ÷ 2
+    V = SVector{D, T}
+    if haskey(bsn_nfo.attractors, attractor_id)
+        push!(bsn_nfo.attractors[attractor_id], V(u))
+    else
+        # initialize container for new attractor
+        bsn_nfo.attractors[attractor_id] = StateSpaceSet([V(u)])
+    end
+end
+
+# Notice that seting a basin index to 0 _deletes the index_ if the
+# array is a `SparseArray`, see the source code file!
+function cleanup_visited_cells!(bsn_nfo::BasinsInfo)
+    old_label = bsn_nfo.visited_cell_label
+    while !isempty(bsn_nfo.visited_list)
+        ind = pop!(bsn_nfo.visited_list)
+        if bsn_nfo.basins[ind] == old_label
+            bsn_nfo.basins[ind] = 0 # 0 is the unvisited label / empty label
+        end
+    end
+end
+
+
+function reset_basins_counters!(bsn_nfo::BasinsInfo)
+    bsn_nfo.consecutive_match = 0
+    bsn_nfo.consecutive_lost = 0
+    bsn_nfo.prev_label = 0
+    bsn_nfo.state = :att_search
+end
+
+function update_fsm_state!(bsn_nfo, ic_label)
```

Comment:
This function is very cryptic. I am sure there is a better way to express the transitions from state to state. I will try to improve this part.